File: app/src/main/java/skarlat/dev/ecoproject/fragment/UserFragment.java
package skarlat.dev.ecoproject.fragment;

import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.io.IOException;
import java.io.InputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLConnection;
import java.util.List;
import java.util.Objects;

import skarlat.dev.ecoproject.EcoCard;
import skarlat.dev.ecoproject.R;
import skarlat.dev.ecoproject.User;
import skarlat.dev.ecoproject.adapter.CategoryAdapter;
import skarlat.dev.ecoproject.App;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link UserFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class UserFragment extends Fragment {
	// TODO: Rename parameter arguments, choose names that match
	// the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
	private static final String ARG_PARAM1 = "param1";
	private static final String ARG_PARAM2 = "param2";
	public static UserFragment userFragment;
	public static String userPhotoUrl;
	public static String userName;
	
	// TODO: Rename and change types of parameters
	private String mParam1;
	private String mParam2;
	
	private List<EcoCard> cards;
	
	public UserFragment() {
		// Required empty public constructor
	}
	
	/**
	 * Use this factory method to create a new instance of
	 * this fragment using the provided parameters.
	 *
	 * @param param1 Parameter 1.
	 * @param param2 Parameter 2.
	 * @return A new instance of fragment UserFragment.
	 */
	// TODO: Rename and change types and number of parameters
	public static UserFragment newInstance(int page) {
		Bundle args = new Bundle();
		args.putInt("UserFragment", page);
		UserFragment fragment = new UserFragment();
		fragment.setArguments(args);
		userFragment = fragment;
		return fragment;
	}
	
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		if (getArguments() != null) {
			mParam1 = getArguments().getString(ARG_PARAM1);
			mParam2 = getArguments().getString(ARG_PARAM2);
		}
	}
	ImageView imageView;
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		// Inflate the layout for this fragment
		View view = inflater.inflate(R.layout.fragment_user, container, false);
		RecyclerView recyclerView = view.findViewById(R.id.cards_by_category);
		TextView textView = view.findViewById(R.id.user_name);
		
		imageView = view.findViewById(R.id.profile_image);
		textView.setText(User.currentUser.name);
		Runnable runnable = new Runnable() {
			@Override
			public void run() {
				showUserAvatar();
			}
		};
		
		cards = App.getDatabase().cardsDao().getAll();
		
		FloatingActionButton fab = view.findViewById(R.id.pressBackFromFragment);
		fab.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				Objects.requireNonNull(getActivity()).onBackPressed();
			onDestroy();}
		});
		// @TODO: Заменить заполнение листа с ипользованием БД
		
		CategoryAdapter adapter = new CategoryAdapter(getContext(), cards);
		DividerItemDecoration itemDecoration = new DividerItemDecoration(getContext(), RecyclerView.HORIZONTAL);
		itemDecoration.setDrawable(getResources().getDrawable(R.drawable.divider_category));
		recyclerView.addItemDecoration(itemDecoration);
		recyclerView.setAdapter(adapter);
		
		return view;
	}
	
	private void showUserAvatar(){
		try {
			URL photoURL = new URL(userPhotoUrl);
			URLConnection urlConnection = photoURL.openConnection();
			InputStream inputStream = urlConnection.getInputStream();
			imageView.setImageDrawable(new BitmapDrawable(getResources(),inputStream));
		} catch (MalformedURLException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}

File: app/src/main/java/skarlat/dev/ecoproject/fragment/UserFragment.kt
package skarlat.dev.ecoproject.fragment

import android.graphics.drawable.BitmapDrawable
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import skarlat.dev.ecoproject.ContentNavigationDirections
import skarlat.dev.ecoproject.EcoTipsApp
import skarlat.dev.ecoproject.R
import skarlat.dev.ecoproject.adapter.CategoryAdapter
import skarlat.dev.ecoproject.databinding.FragmentUserBinding
import skarlat.dev.ecoproject.includes.dataclass.EcoCard
import skarlat.dev.ecoproject.network.FirebaseAPI
import java.io.IOException
import java.net.MalformedURLException
import java.net.URL

class UserFragment : BaseFragment<FragmentUserBinding>() {
    private var cards: List<EcoCard>? = null

    var imageView: ImageView? = null

    override fun inflateBinding(): ViewBinding {
        return FragmentUserBinding.inflate(layoutInflater)
    }

    private fun initUI() {
        cards = EcoTipsApp.getDatabase().cardsDao().all
        binding.pressBackFromFragment.setOnClickListener {
            requireActivity().onBackPressed()
        }
        // @TODO: Заменить заполнение листа с ипользованием БД
        val adapter = CategoryAdapter(context, cards)
        val itemDecoration = DividerItemDecoration(context, RecyclerView.HORIZONTAL)
        itemDecoration.setDrawable(resources.getDrawable(R.drawable.divider_category))
        binding.cardsByCategory.addItemDecoration(itemDecoration)
        binding.cardsByCategory.adapter = adapter
        binding.logoutButton.setOnClickListener {
            viewLifecycleOwner.lifecycleScope.launch {
                EcoTipsApp.appComponent.getAuthManager().logout()
                findNavController().navigate(ContentNavigationDirections.globalActionToMain())
                requireActivity().finishAfterTransition()
            }
        }
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.openSettingsButton.setOnClickListener {
//            fragmentManager!!.beginTransaction().add(R.id.home_layout, ProfileSettingsFragment.newInstance()).commit()
        }
        initUI()
    }

    override fun onStart() {
        super.onStart()
        val appCache = EcoTipsApp.appComponent.getAppCache()
        FirebaseAPI.getUsersCount { appCache.setCount(it.toInt()) }
        viewLifecycleOwner.lifecycleScope.launchWhenStarted {
            launch {
                appCache.userProgressFlow.collectLatest {
                    binding.percentProgress.text = "${it}%"
                    binding.totalProgressBar.progress = it
                }
            }
            launch {
                appCache.userFlow.collectLatest {
                    binding.userName.text = it.name
                }
            }
            launch {
                appCache.userCountFlow.collectLatest {
                    binding.countUsers.text = getString(R.string.count_users_format, it)
                }
            }
        }
    }

    private fun showUserAvatar() {
        // TODO Use glide
        try {
            val photoURL = URL(userPhotoUrl)
            val urlConnection = photoURL.openConnection()
            val inputStream = urlConnection.getInputStream()
            imageView!!.setImageDrawable(BitmapDrawable(resources, inputStream))
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    companion object {
        var userPhotoUrl: String? = null
    }
}